src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -390,17 +390,20 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
             }
 
             if (_buckets == null) Initialize(0);
-            int hashCode = _comparer.GetHashCode(key) & 0x7FFFFFFF;
+            IEqualityComparer<TKey> comparer = _comparer;
+            int hashCode = comparer.GetHashCode(key) & 0x7FFFFFFF;
             int targetBucket = hashCode % _buckets.Length;
             int collisionCount = 0;
 
-            for (int i = _buckets[targetBucket]; i >= 0; i = _entries[i].next)
+            int i = _buckets[targetBucket];
+            Entry[] entries = _entries;
+            while (i >= 0)
```

Comment:
In what way is **while** better than a **for** loop?
In general (so not specific to the changes here) we're going from foreach - > for loop - > while (with locals).
What's next? :p

Having such manually optimized code in this repo is ok, but I honestly hope the JIT can be improved further so that (performant) user code can remain simple and readable.  And yes, I know there's work in progress (eg. https://github.com/dotnet/coreclr/issues/15130)

Reply:
> In what way is while better than a for loop?

When the for loop is not really a for loop, but a while loop in disguise. 

The "increment" portion `; i = _entries[i].next)` causes an additional bounds check as a `for` but not as a `while`

Reply:
While you're at try `while ((uint)i < (uint)entries.Length)`, usually this doesn't work in loops but I've seen a case where it did, and I think it was a `while` too.

Reply:
Though I probably need to take a closer look at this, it's not readily obvious why `while` is different than `for` in this case. Does the difference persist if you move `int i = _buckets[targetBucket];` after `Entry[] entries = _entries;`?

Reply:
Seems a bit sensitive

PR
```
Total bytes of diff: -3687 (-0.11% of base)
(5 range checks, bytes of code 565)
-249 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
(6 range checks, bytes of code 894)
-3 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this 
```
PR + for loop (local var)
```
Total bytes of diff: -2860 (-0.08% of base)
(6 range checks, bytes of code 594)
-162 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
(7 range checks, bytes of code 902)
+5 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this
```
PR + for loop (local var) + `(uint)i < (uint)entries.Length` condition
```
Total bytes of diff: -3093 (-0.09% of base)
(6 range checks, bytes of code 579)
-207 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
(7 range checks, bytes of code 891)
-6 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this
```
PR + switching `Entry[] entries =` with `int i =`
```
Total bytes of diff: -3659 (-0.11% of base)
(5 range checks, bytes of code 569)
-237 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
(6 range checks, bytes of code 894)
-3 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this
```
PR + `while ((uint)i < (uint)entries.Length)`
```
Total bytes of diff: -3607 (-0.11% of base)
(5 range checks, bytes of code 573)
-225 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
(6 range checks, bytes of code 870)
-27 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this 
```
PR + `while ((uint)i < (uint)entries.Length)` + switching `Entry[] entries =` with `int i =`
```
Total bytes of diff: -3594 (-0.11% of base)
(5 range checks, bytes of code 576)
-216 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
(6 range checks, bytes of code 870)
-27 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this 
```

5/6 range checks in this PR still seems a bit high

Reply:
A simpler one to look at might be https://github.com/dotnet/coreclr/pull/15460

Changing to `(uint)i < (uint)entries.Length` doesn't skip bounds check there either

Reply:
Added issue https://github.com/dotnet/coreclr/issues/15476